packages/test/test-utils/src/testObjectProvider.ts

```diff
@@ -474,10 +591,339 @@ export class TestObjectProvider implements ITestObjectProvider {
 		return waitContainerToCatchUp_original(container);
 	}
 
-	updateDocumentId(resolvedUrl: IResolvedUrl | undefined) {
+	/**
+	 * {@inheritDoc ITestObjectProvider.updateDocumentId}
+	 */
+	public updateDocumentId(resolvedUrl: IResolvedUrl | undefined) {
 		this._documentIdStrategy.update(resolvedUrl);
 	}
 
+	/**
+	 * {@inheritDoc ITestObjectProvider.resetLoaderContainerTracker}
+	 */
+	public resetLoaderContainerTracker(syncSummarizerClients: boolean = false) {
+		this._loaderContainerTracker.reset();
+		this._loaderContainerTracker = new LoaderContainerTracker(syncSummarizerClients);
+	}
+}
+
+/**
+ * Implements {@link ITestObjectProvider}, but uses different versions to create and load containers.
+ *
+ * @internal
+ */
+export class TestObjectProviderWithVersionedLoad implements ITestObjectProvider {
+	/**
+	 * {@inheritDoc ITestObjectProvider."type"}
+	 */
+	public readonly type = "TestObjectProviderWithVersionedLoad";
+	private _loaderContainerTracker = new LoaderContainerTracker();
+	private _logger: EventAndErrorTrackingLogger | undefined;
+	private readonly _documentIdStrategy: IDocumentIdStrategy;
+	private _documentServiceFactory: IDocumentServiceFactory | undefined;
+	private _urlResolver: IUrlResolver | undefined;
+	// Since documentId doesn't change we can only create/make one container. Call the load functions instead.
+	private _documentCreated = false;
+	// `_loadCount` is used to alternate which version we load the next container with.
```

Comment:
Use a doc comment here so it shows up in intellisense.  Also since it increments on creating a loader (not on loading), should it be called `_loaderCount`?

Based on the PR description, I think this alternating behavior is kind of a "shotgun" approach, that some tests use more than one loader and so will happen to load different versions up.  Whereas the "opt-in" approach you mention might look more like an explicit API call to get the old vs. new version?

One thing to consider - this count (and the effects it has) are maintained at the provider level, but some of the things they control probably want to be inspected on a per-Loader basis.  This could introduce some fragility, e.g. checking provider.driver is only accurate for the Loader I'm holding if no other Loader was created in the meantime.

So maybe a couple recommendations here:
1. If there is a different design we may consider later (e.g. the opt-in you mention) then perhaps hint at that in here somewhere with a link to work item.
2. Consider whether the fragility I mention above is a risk - if so, maybe at least document what restrictions it imposes on callers to avoid tripping up (e.g. query all properties you're interested in before creating your next Loader).

Reply:
Yeah, I think long-term we want to avoid the "shotgun" approach. As I mentioned in the PR description, I think at this moment it's probably still the best course of action, but I can open up a follow up item to look into it further